frontend-react/docs/proposals/0003-uswds-react-components.md

```diff
@@ -0,0 +1,251 @@
+- Ideal for stylistic changes to Trussworks React components. Can simply use the scoped `.module` syntax to create new dynamic styles.
+- No need to add styles to our convoluted overrides file `_uswds_overrides.scss` anymore.
+
+#### Cons
+- Not ideal for complex functional extensions of React components.
+- Can get convoluted quickly since we're essentially wrapping a wrapper.
+- Potentially easy breaking changes as we'd be two levels from USWDS.
+- Still need to manually override USWDS styles on occasion using `!important`.
+
+Sample `<RSTable/>` component with a `stickyHeader` feature:
+
+`RSTable.jsx`:
+
+```js
+import { Table } from "@trussworks/react-uswds";
+
+import styles from "./RSTable.module.css";
+
+export const RSTable = ({ stickyHeader, children, ...props }) => {
+    return (
+        <div className={stickyHeader && styles.stickyHeader}>
+            <Table {...props}>{children}</Table>
+        </div>
+    );
+};
+```
+
+`RSTable.module.css`:
+
+```css
+.stickyHeader thead th {
+    position: sticky;
+    top: 0;
+}
+```
+
+---
+
+<a name="option-two"></a>
+### Option 2: Utilize USWDS Base Library Only
+
+#### Description
+
+Utilize the raw [USWDS library](https://github.com/uswds/uswds) which would remove a level of abstraction to our components. We'd then wrap their components and create a React component that interacts directly with their code which will allow us full customization of props and functionality. CSS will still need to be overridden at times, but can be component-scoped.
+
+#### Pros
+
+- 1:1 relationship with USWDS which means we can upgrade our dependencies without having to rely on Trussworks upgrading theirs.
+- Removes a layer of abstraction so we can create React components with naming conventions and styling that more closely match our designs and syntax.
+- DOM structure will be exposed so we can more easily understand what's going on under the hood which will also help with writing better tests and writing more complex, yet clear, components.
+
+#### Cons
+- Potential breaking changes as we'd be building on top of USWDS.
+- Still need to manually override USWDS styles on occasion using `!important`.
+- Not ideal for very complicated features, like some of the more advanced ideas we had for a ReportStream table.
+
+`RSTable.jsx`:
+
+```js
+import styles from "./RSTable.module.css";
+import classnames from "classnames";
+
+export const RSTable = ({ props, children }) => {
+    // We could easily make a master hash file
+    // with all USWDS CSS classes instead
+    const USWDSTableClasses = {
+      borderless: 'usa-table--borderless',
+      compact: 'usa-table--compact',
+      striped: 'usa-table--striped',
+      stickyHeader: 'rs-table--sticky-header',
+    }
+    
+    // The classnames library recursively flattens arrays
+    const classes = classnames("usa-table", Object.keys(props).map((class) => USWDSTableClasses[class]));
```

Comment:
I know this is a hypothetical component, but I feel like it'd be clearer to pass this as an object to `classnames`:

```jsx
const classes = classnames("usa-table", {
    "usa-table--borderless": borderless,
    "usa-table--compact": compact,
    ...
});
```

This'd still be usable if we had a hash or enum of available styles provided by the USWDS too:

```jsx
const classes = classnames("usa-table", {
    [UswdsStyle.Borderless]: borderless,
    [UswdsStyle.Compact]: compact,
    ...
});
```

We could also play around with the concept of `variant`s as well.  But just a thought! 🤷

Reply:
I'm all for being explicit! So you're saying to forego the iteration with `Object.keys(props)` and just being explicit with the `USWDSTableClasses` <-> 'className` mapping? Totally cool with that.

Reply:
Yeah, I think just being explicit helps out here because it's slightly less mental overhead to process :)